diff --git a/kernel/msm-4.19/techpack/display/oplus/oplus_display_panel.c b/kernel/msm-4.19/techpack/display/oplus/oplus_display_panel.c
index 45d62e6120ce..1dc3f25689d4 100755
--- a/kernel/msm-4.19/techpack/display/oplus/oplus_display_panel.c
+++ b/kernel/msm-4.19/techpack/display/oplus/oplus_display_panel.c
@@ -477,2 +477,2 @@
-void __exit oplus_display_panel_exit()
+void __exit oplus_display_panel_exit(void)
 {
diff --git a/vendor/oplus/kernel/system/uboot_log/uboot_log.c b/vendor/oplus/kernel/system/uboot_log/uboot_log.c
index b6e45c3be1ca..4a1d173e7c3b 100755
--- a/vendor/oplus/kernel/system/uboot_log/uboot_log.c
+++ b/vendor/oplus/kernel/system/uboot_log/uboot_log.c
@@ -170,2 +170,2 @@
-int ubootback_thread_fn()
+int ubootback_thread_fn(void *data)
 {
